python/lsst/meas/modelfit/optimizer/optimizer.cc

```diff
@@ -0,0 +1,207 @@
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionInitialSize);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowReductionRatio);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowStepFraction);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionGrowFactor);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionShrinkReductionRatio);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionShrinkFactor);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, trustRegionSolverTolerance);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, maxInnerIterations);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, maxOuterIterations);
+    LSST_DECLARE_CONTROL_FIELD(cls, OptimizerControl, doSaveIterations);
+    cls.def(py::init<>());
+    return cls;
+}
+
+PyOptimizerHistoryRecorder declareOptimizerHistoryRecorder(py::module & mod) {
+    PyOptimizerHistoryRecorder cls(mod, "OptimizerHistoryRecorder");
+    cls.def(py::init<afw::table::Schema &, std::shared_ptr<Model>, bool>(),
+            "schema"_a, "model"_a, "doRecordDerivatives"_a);
+    cls.def(py::init<afw::table::Schema const &>(), "schema"_a);
+    cls.def("apply", &OptimizerHistoryRecorder::apply,
+            "outerIterCount"_a, "innerIterCount"_a, "history"_a, "optimizer"_a);
+    cls.def(
+        "unpackDerivatives",
+        (void (OptimizerHistoryRecorder::*)(
+            ndarray::Array<Scalar const,1,1> const &,
+            ndarray::Array<Scalar,1,1> const &,
+            ndarray::Array<Scalar,2,2> const &
+        ) const) &OptimizerHistoryRecorder::unpackDerivatives,
+        "nested"_a, "gradient"_a, "hessian"_a
+    );
+    cls.def(
+        "unpackDerivatives",
+        (void (OptimizerHistoryRecorder::*)(
+            afw::table::BaseRecord const &,
+            ndarray::Array<Scalar,1,1> const &,
+            ndarray::Array<Scalar,2,2> const &
+        ) const) &OptimizerHistoryRecorder::unpackDerivatives,
+        "record"_a, "gradient"_a, "hessian"_a
+    );
+    cls.def("fillObjectiveModelGrid", &OptimizerHistoryRecorder::fillObjectiveModelGrid,
+            "record"_a, "parameters"_a, "output"_a);
+    cls.def_readonly("outer", &OptimizerHistoryRecorder::outer);
+    cls.def_readonly("inner", &OptimizerHistoryRecorder::inner);
+    cls.def_readonly("state", &OptimizerHistoryRecorder::state);
+    cls.def_readonly("objective", &OptimizerHistoryRecorder::objective);
+    cls.def_readonly("prior", &OptimizerHistoryRecorder::prior);
+    cls.def_readonly("trust", &OptimizerHistoryRecorder::trust);
+    cls.def_readonly("parameters", &OptimizerHistoryRecorder::parameters);
+    cls.def_readonly("derivatives", &OptimizerHistoryRecorder::derivatives);
+    return cls;
+}
+
+PyOptimizer declareOptimizer(py::module & mod) {
+    PyOptimizer cls(mod, "Optimizer");
+    // StateFlags enum is used as bitflag, so we wrap values as int class attributes.
```

Comment:
You are saying that in general if an enum is used to define a bitmask that the bitmask type should be int-like instead of the enum class? Interesting. I like the type safety of having to use named constants, but it is, admittedly, a headache when specifying a value in a config (as for all enums, due to problems with the config design) and I it is difficult to check when persisting and unpersisting. Perhaps this is a case where pybind11 is unhelpful in its type safety; being able to define such objects using enum seems natural but makes life hard for us.